core/src/main/java/com/novoda/noplayer/internal/exoplayer/mediasource/MediaSourceFactory.java

```diff
@@ -43,19 +42,18 @@ public MediaSourceFactory(Context context,
 
     public MediaSource create(Options options,
                               Uri uri,
-                              MediaSourceEventListener mediaSourceEventListener,
-                              DefaultBandwidthMeter bandwidthMeter,
-                              Optional<AdsLoader> advertsLoader) {
+                              ExoPlayerListener exoPlayerListener,
+                              DefaultBandwidthMeter bandwidthMeter) {
         DefaultDataSourceFactory defaultDataSourceFactory = createDataSourceFactory(bandwidthMeter);
 
         MediaSource contentMediaSource = getMediaSourceFor(options, uri, defaultDataSourceFactory);
 
-        if (advertsLoader.isPresent()) {
-            AdsMediaSource adsMediaSource = new AdsMediaSource(contentMediaSource, defaultDataSourceFactory, advertsLoader.get(), null);
-            adsMediaSource.addEventListener(handler, mediaSourceEventListener);
+        if (exoPlayerListener.canLoadAdverts()) {
+            AdsMediaSource adsMediaSource = new AdsMediaSource(contentMediaSource, defaultDataSourceFactory, exoPlayerListener, null);
```

Comment:
As far I understand the `AdsMediaSource` here is decorating the `contentMediasource`, but I don't fully grasp why you need to add listener when creating one.  Shouldn't be responsibility of the consumer deciding whether they are interested in adverts (so to wrap the original source) and then subscribing to those additional events too?